packages/shared/src/session/deferred.ts

```diff
@@ -0,0 +1,6 @@
+import { createDeferred } from "../async/createDeferred";
+import { AuthInfo } from "../authentication/types";
+import { PackageInfo } from "./types";
+
+export const deferredAuthInfo = createDeferred<AuthInfo | null>();
```

Comment:
This is slightly incorrect. It should be this
```suggestion
export const deferredAuthInfo = createDeferred<AuthInfo | undefined>();
```
That's because `packages/shared/src/session/initializeSession.ts` can resolve it with `undefined` and not `null`.

This has accidentally been allowed because `deferred.resolve` has an optional argument. We could make optionally optional (when the `Type` allows `undefined`) but I think it's not worth the added complexity like this:
```ts
type Foo<T> = {
  method: (...args: undefined extends T ? [a?: T] : [a: T]) => void;
};

declare const foo1: Foo<{ bar: string } | null>;
// @ts-expect-error
foo1.method();
// @ts-expect-error
foo1.method(undefined);

declare const foo2: Foo<{ bar: string } | undefined>;
foo2.method();
foo2.method(undefined);
```

So I'd go with a patch like this:
```diff
diff --git a/packages/shared/src/async/createDeferred.ts b/packages/shared/src/async/createDeferred.ts
index 8558bd50..3fffb90e 100644
--- a/packages/shared/src/async/createDeferred.ts
+++ b/packages/shared/src/async/createDeferred.ts
@@ -11,10 +11,10 @@ export interface Deferred<Type, Data = undefined> {
   debugLabel: string | undefined;
   promise: Promise<Type>;
   rejection: Error | undefined;
-  reject(error: Error): void;
-  rejectIfPending(error: Error): void;
-  resolve(value?: Type): void;
-  resolveIfPending(value?: Type): void;
+  reject: (error: Error) => void;
+  rejectIfPending: (error: Error) => void;
+  resolve: (value: Type) => void;
+  resolveIfPending: (value: Type) => void;
   resolution: Type | undefined;
   status: StatusPending | StatusRejected | StatusResolved;
 }
```

Reply:
Hm, we actually often call `deferred.resolve()` so to avoid problems we might want to use the complicated version outlined above 😢

Reply:
I'm not totally sure I understand what you're proposing in the "complicated option" above, but if it's okay I'll fix the incorrect type (`createDeferred<AuthInfo | null>()` -> `createDeferred<AuthInfo | undefined>()`) and we can address the deferred API signatures in a follow up.